include/dr/mhp/containers/distributed_dense_matrix.hpp

```diff
@@ -0,0 +1,174 @@
+
+  dm_tile(T * data, key_type shape, size_t rank) 
+    : data_(data), rank_(rank), shape_(shape) {};
+
+  bool is_local() { return rank_ == default_comm().rank(); }
+
+  size_t rank() { return rank_; }
+  key_type shape() { return shape_; }
+private:
+  T * data_ = nullptr;
+  size_t rank_;
+  key_type shape_;
+
+};
+
+template <typename T> class dm_row_iterator : 
+  public std::iterator<std::forward_iterator_tag, T> {
+public:
+  dm_row_iterator(T* p, dm_tile<T> & tile) : p_(p), tile_(tile) {};
+
+  dm_row_iterator& operator++() { p_++; return *this; };
+  bool operator==(dm_row_iterator other) const { return p_ == other.p_; }
+  bool operator!=(dm_row_iterator other) const { return p_ != other.p_; }
+  T & operator*() { return *p_; }
+
+private:
+  T* p_;
+  dm_tile<T> & tile_;
+};
+
+template <typename T> class dm_row_view {
+  using iterator = dm_row_iterator<T>;
+public:
+  dm_row_view(T * ptr, dm_tile<T> & tile, size_t size) : data_(ptr), tileref_(tile), size_(size) {};
+
+  iterator begin() { return  iterator(data_, tileref_); }
+  iterator end() { return  iterator(data_ + size_, tileref_); }
+private:
+  T * data_ = nullptr;
+  dm_tile<T> & tileref_;
+  size_t size_ = 0;
+};
+
+template <typename T, typename Allocator = std::allocator<T>> 
+class distributed_dense_matrix {
+public:
+  using size_type = std::size_t;
+  using difference_type = std::ptrdiff_t;
+
+  // using value_type = mhp::matrix_entry<T>;
+
+  using key_type = mhp::index<>;
+
+  using iterator =
+      lib::iterator_adaptor<distributed_dense_matrix<T>>;
+
+  size_type size() const noexcept { return shape()[0] * shape()[1]; }
+
+  key_type shape() const noexcept { return shape_; }
+
+    
+  std::vector<dm_row_view<T>> local_rows() {
+    std::vector<dm_row_view<T>> row_view_;
+
+    for(auto _titr = tiles_.begin(); _titr != tiles_.end(); ++_titr) {
+      if ((*_titr).is_local()) {
+        for (size_t s = 0; s < (*_titr).shape()[0]; s++) {
+          row_view_.emplace_back(data_ + s * (*_titr).shape()[1], *_titr, (*_titr).shape()[1]);
+        }
+      }
+    };
+    return row_view_;
+  }
+
+  std::vector<dm_row_view<T>> rows() {
+    std::vector<dm_row_view<T>> row_view_;
+
+    for(auto _titr = tiles_.begin(); _titr != tiles_.end(); ++_titr) {
+      for (size_t s = 0; s < (*_titr).shape()[0]; s++) {
+        T * _dataptr = (*_titr).is_local() ? (data_ + s * (*_titr).shape()[1]) : nullptr;
+        row_view_.emplace_back(_dataptr, *_titr, (*_titr).shape()[1]);
+      }
+    };
+    return row_view_;
+  }
+
+
+
+  distributed_dense_matrix(key_type shape, lib::halo_bounds hb = lib::halo_bounds(), Allocator allocator = Allocator())
+      : shape_(shape), // partition_(new mhp::block_cyclic()),
+      grid_shape_(default_comm().size(), 1),
+      tile_shape_(( shape_[0] + default_comm().size() - 1)/default_comm().size(), shape_[1]),
+      segment_size_( std::max({ tile_shape_[0] * tile_shape_[1], hb.prev * tile_shape_[1], hb.next * tile_shape_[1] } )),
```

Comment:
hb.prev affects tile_shape. A tile must be at least as big as a halo in all dimensions. halo needs to be extended to 2 dimensions. Since you don't have any of that, you should delete the halo and add it back in when you are ready to fully implement halo.